src/iarc7_motion/linear_motion_profile_generator.py

```diff
@@ -0,0 +1,209 @@
+        v_delta = v_desired - v_start
+
+        # Assign a direction to the target acceleration
+        a_target = self._TARGET_ACCEL * v_delta / np.linalg.norm(v_delta)
+
+        # Find the time required to accelerate to the desired velocity
+        acceleration_time = min(np.linalg.norm(v_delta) / self._TARGET_ACCEL, self._PLAN_DURATION)
+
+        # Use the rest of the profile duration to hold the velocity
+        steady_velocity_time = self._PLAN_DURATION - acceleration_time
+
+        # Calculate the number of discrete steps spent in each state
+        accel_steps = np.floor(acceleration_time / self._PROFILE_TIMESTEP)
+        vel_steps = np.floor(steady_velocity_time / self._PROFILE_TIMESTEP)
+
+        # Initialize the velocities array with the starting velocity
+        velocities = [v_start]
+
+        # Generate velocities corresponding to the acceleration period
+        if accel_steps > 0:
+            accel_times = (np.linspace(0.0, accel_steps*self._PROFILE_TIMESTEP, accel_steps)
+                     * accel_steps
+                     * self._PROFILE_TIMESTEP)
+            velocities.extend([a_target * accel_time + v_start for accel_time in accel_times])
+
+        # Add the velocities for the steady velocity period
+        velocities.extend([v_desired for i in range(0, int(vel_steps))])
+
+        velocities = np.array(velocities)
+
+        # Differentiate velocities to get the accelerations
+        # This results in an array one element shorter than
+        # the velocities array
+        accelerations = np.diff(velocities, axis=0) / self._PROFILE_TIMESTEP
+
+        # Integrate the velocities to get the position deltas
+        # This results in an array the same length as the velocities array
+        # But the positions correspond to the velocities one index higher than
+        # the given positio index
+        positions = (np.cumsum(velocities, axis=0) * self._PROFILE_TIMESTEP
+                     + p_start)
+
+        plan = MotionPointStampedArray()
+        pose_only_plan = Path()
+        pose_only_plan.header.stamp = rospy.Time.now()
+        pose_only_plan.header.frame_id = 'map'
+
+        # Fill out the first motion point since it follows different
+        # rules than the rest
+        motion_point = MotionPointStamped()
+        motion_point.header.stamp = start_point.header.stamp
+        np_to_msg(accelerations[0], motion_point.motion_point.accel.linear)
+        np_to_msg(v_start, motion_point.motion_point.twist.linear)
+        np_to_msg(p_start, motion_point.motion_point.pose.position)
+
+        plan.motion_points.append(motion_point)
+        pose = PoseStamped()
+        pose.header.stamp = motion_point.header.stamp
+        pose.pose = motion_point.motion_point.pose
+        pose_only_plan.poses.append(pose)
+
+        # Generate the motion profile for all the remaining velocities and accelerations
+        for i in range(1, velocities.shape[0]-1):
```

Comment:
Is there a reason we're throwing out the last point in `velocities`?

Reply:
Yes because we don't have an acceleration for the last time.